resources/2024-06-resource-extensibility-feature-spec.md

```diff
@@ -0,0 +1,452 @@
+
+### Top level goals
+
+1. Enable users to author their applications with custom resource types in Radius without having to write Go code to integrate with Radius components and deploy their application seamlessly.
+2. Automatically enable Radius features such as Recipes, Connections, App graph, rad CLI for custom resource types.
+3. Enable users to contribute and open-source the custom resource types and recipes to the community. This will accelerate the adoption of Radius across the cloud native community.
+4. Radius uses the resource extensibility framework to refactor its already built in resource types.
+
+### Non-goals (out of scope)
+
+We will not cover the design and user experience for "Recipes for any resource" in this document to keep the scope of the work simple.
+
+
+## User profile and challenges
+
+**Enterprises**: Platform engineering teams or operations in enterprises focus on streamlining the developer-experience for their organization by defining a set of recommended practices for application teams and provide self-service capabilities for application teams/developers to use. Radius aids the platform engineering efforts with the help of Recipes where the platform engineers or operators define the infrastructure-as-code templates to create resources on demand when application is deployed. One of the major challenges that exists with the Recipes today is the limited number of resources supported. Radius doesn’t provide an extensibility model for users to bring their custom resources, define Recipes and deploy them.
+
+**Open-source community**: Building a sustainable open-source community is crucial to the success of any open-source project, including Radius. We need to cultivate an ecosystem for the open-source community to thrive, innovate and produce high quality work. Users from the community are motivated to contribute for different reasons:
+    -  They want to use the project at their work and need a feature, 
+    -  The project triggered their interests in the cloud native landscape and want to use the technology.
+    -  They want to code in a particular language E.g. :Go
+    -  They want to integrate their existing tools and technologies with the project
+
+Today Radius enables users to get started on contributing to Radius with good-first-issues but doesn’t have a model to promote contributions further. Since the beginning of its open source launch, Dapr created the components repository and invited users to build their own components to unblock their scenarios and in turn folks contributed to the project. Learning from Dapr, Radius needs to have the extensibility points defined for contributors to interact with the project so that they can build out the things that they need to enable their scenarios and in turn the community gets benefitted with those contributions.
```

Comment:
```suggestion
Today Radius enables users to get started on contributing to Radius with good-first-issues but doesn’t have a model to promote contributions further. Since the beginning of its open source launch, [Dapr](www.dapr.io) created the components repository and invited users to build their own components to unblock their scenarios and in turn folks contributed to the project. Learning from Dapr, Radius needs to have the extensibility points defined for contributors to interact with the project so that they can build out the things that they need to enable their scenarios and in turn the community gets benefitted with those contributions.
```